accepted/00000-minion-selective-action-blocking.md

```diff
@@ -0,0 +1,93 @@
+
+
+# Detailed design
+[design]: #detailed-design
+
+Traditional clients can be locked by the user. When locked, only certain action types are allowed. Actions that change the state of a system are blocked (package install, upgrade, apply highstate, etc) while read-only actions are allowed (hardware refresh, package profile refresh, etc).
+
+The same should be possible for Salt minions. However, in the case of Salt minions, additional flexibility is needed to allow blocking some actions depending on the the products installed on the minion. This is needed to accommodate CaaSP and maybe other clustering products in the future.
+
+To prevent the user from accidentally executing forbidden actions, the selective blocking must work both at the level of the Uyuni UI/XML-RPC API and at the level of the Salt command line.
+
+## Salt level selective action blocking
+
+The Salt blackout mechanism works by defining a special pillar called `minion_blackout` to tell the minion that it must reject all remote executions. Exceptions can be configured using the `minion_blackout_whitelist` pillar. This accepts a list of Salt functions that are allowed during blackout, e.g. `test.ping`, `state.apply`, etc.
+
+However, in Uyuni there are multiple actions that use `state.appply` under the hood. Some actions can change the state of the system while others are read-only but they all use `state.apply`. If `minion_blackout_whitelist` is used to allow `state.apply` then all actions that use this function are allowed.
+
+The blackout mechanism must be enhanced to allow a more fine grained filtering. One approach would be to check the metadata that can be attached to a Salt call. 
+
+A metadata whitelist would be defined. If the Salt call contains metadata that matches the whitelisted metadata then the call would be allowed. 
+
+```yaml
+minion_blackout: True
+minion_blackout_metadata_whitelist:
+    minion-action-type:
+    - packages.refresh_list
+    - hardware.refresh_list
+```
+
+## Uyuni level selective action blocking
+
+For traditional clients the action types that are allowed when a system is locked are configured in the db. The configuration is global and applies to all clients indiscriminately.
+
+In the case of Salt minions, the same approach will be used. The action types that are allowed when a system is locked will be stored in the db.
+
+Additionally, in order to make the approach more flexible and to accommodate CaaSP a new object called a "locking profile" will be introduced.
+
+The locking profile will contain a list of permitted actions. A profile can be reused across multiple minions. A minion will have only one locking profile assigned when locked. 
```

Comment:
Would it be possible a scenario where we would need two or more locking profiles associated with a minion?
In that case, what would the resulting operation for the locking profiles, union?

Reply:
To keep things simple for now I would say to have only one locking profile per minion. This can be easily changed if such a need appears in the future.